generated/nidaqmx/_grpc_interpreter.py

```diff
@@ -2269,7 +2269,7 @@ def read_analog_f64(
                 task=task, num_samps_per_chan=num_samps_per_chan,
                 timeout=timeout, fill_mode_raw=fill_mode,
                 array_size_in_samps=read_array.size))
-        read_array[:] = response.read_array
+        read_array[:] = numpy.reshape(response.read_array, read_array.shape)
```

Comment:
Using reshape on the right hand side of the assignment makes an unnecessary copy because it must convert the RHS to a numpy array:
```
>>> import numpy
>>> x=[1,2,3,4]
>>> y = numpy.reshape(x, (2,2))
>>> y
array([[1, 2],
       [3, 4]])
>>> x[0]=10
>>> y
array([[1, 2],
       [3, 4]])
```

The left hand side of the assignment is already a numpy array, so if you use reshape on the LHS then it will create a view object for the existing array and it will not make a copy.
```
>>> import numpy
>>> x = numpy.array([[1,2],[3,4]])
>>> x
array([[1, 2],
       [3, 4]])
>>> numpy.reshape(x, (4,))[:] = [5,6,7,8]
>>> x
array([[5, 6],
       [7, 8]])
```

Reply:
Update: there is a simpler solution, `read_array.flat`
https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flat.html

```
>>> x.flat = [10,11,12,13]
>>> x
array([[10, 11],
       [12, 13]])
```

Reply:
Another problem: `read_array` may not be the same size as `response.read_array`. See https://github.com/ni/nidaqmx-python/blob/5657d118f80426ab130a6f83de4cca7be366b085/generated/nidaqmx/task.py#L617

This does not have much test coverage, but I think it has some (see line 623)
![image](https://user-images.githubusercontent.com/17914399/236893792-88c0a522-1167-49bc-a276-eaf2385fe187.png)

`numpy.reshape` raises an exception when this happens:
```
>>> numpy.reshape(x, (6,))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 200, in reshape
  File "D:\dev\nidaqmx-python\.venv\lib\site-packages\numpy\core\fromnumeric.py", line 298, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File "D:\dev\nidaqmx-python\.venv\lib\site-packages\numpy\core\fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
ValueError: cannot reshape array of size 4 into shape (6,)
>>> numpy.reshape(x, (2,))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 200, in reshape
  File "D:\dev\nidaqmx-python\.venv\lib\site-packages\numpy\core\fromnumeric.py", line 298, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File "D:\dev\nidaqmx-python\.venv\lib\site-packages\numpy\core\fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
ValueError: cannot reshape array of size 4 into shape (2,)
```

`read_array.flat = ...` doesn't raise an exception, but it performs unnecessary assignments: 
```
>>> import numpy
>>> x = numpy.array([[1,2],[3,4]])
>>> x.flat = [5,6,7]
>>> x
array([[5, 6],
       [7, 5]])
>>> x.flat = [10,11]
>>> x
array([[10, 11],
       [10, 11]])
```

`read_array.flat[:len(response.read_array)] = response.read_array` leaves the extra array elements unmodified when `len(read_array) > len(response.read_array)`:
```
>>> x.flat[:2] = [20,21]
>>> x
array([[20, 21],
       [10, 11]])
```

and it ignores the extra array elements when `len(read_array) < len(response.read_array)`:
```
>>> x.flat[:6] = [30,31,32,33,34,35]
>>> x
array([[30, 31],
       [32, 33]])
```

Reply:
Also: `numpy.copyto` assigns the extra array elements when `len(read_array) > len(response.read_array)` but it raises an exception when `len(read_array) < len(response.read_array)`:
```
>>> numpy.copyto(x, [50,51])
>>> x
array([[50, 51],
       [50, 51]])
>>> numpy.copyto(x, [60,61,62,63,64,65])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 200, in copyto
ValueError: could not broadcast input array from shape (6,) into shape (2,2)
```

Out of these possible solutions, I think think `read_array.flat[:len(response.read_array)] = response.read_array` is the best. To ensure that we aren't losing any array elements, we could put it in a helper function with an assertion or runtime check:
```
def assign_numpy_array(numpy_array, grpc_array):
    assert len(grpc_array) >= numpy_array.size:
    numpy_array.flat[:len(grpc_array)] = grpc_array
```

Reply:
I have created a function in utils to do a runtime check before assigning the array, and if the condition fails it ignore the assignment. Would it be better if we threw warnings in such cases?

Reply:
I looked at what errors and warnings DAQmx defines:
- There is no appropriate DAQmx warning number for this case.
- `DAQmxErrors.WRITE_BUFFER_TOO_SMALL` is appropriate for this case.